plugins/in_statsd/statsd.c

```diff
@@ -0,0 +1,338 @@
+/* -*- Mode: C; tab-width: 4; indent-tabs-mode: nil; c-basic-offset: 4 -*- */
+
+/*  Fluent Bit
+ *  ==========
+ *  Copyright (C) 2019      The Fluent Bit Authors
+ *  Copyright (C) 2015-2018 Treasure Data Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include <fluent-bit/flb_info.h>
+#include <fluent-bit/flb_input.h>
+#include <fluent-bit/flb_utils.h>
+#include <fluent-bit/flb_socket.h>
+#include <fluent-bit/flb_pack.h>
+
+#define MAX_PACKET_SIZE 65536
+#define DEFAULT_LISTEN "0.0.0.0"
+#define DEFAULT_PORT 8125
+
+#define STATSD_TYPE_COUNTER 1
+#define STATSD_TYPE_GAUGE   2
+#define STATSD_TYPE_TIMER   3
+#define STATSD_TYPE_SET     4
+
+struct flb_statsd {
+    char *buf;                         /* buffer */
+    char listen[256];                  /* listening address (RFC-2181) */
+    char port[6];                      /* listening port (RFC-793) */
+    flb_sockfd_t server_fd;            /* server socket */
+    flb_pipefd_t coll_fd;              /* server handler */
+    struct flb_input_instance *i_ins;  /* input instance */
+};
+
+/*
+ * The "statsd_message" represents a single line in UDP packet.
+ * It's just a bunch of pointers to ephemeral buffer.
+ */
+struct statsd_message {
+    char *bucket;
+    int bucket_len;
+    char *value;
+    int value_len;
+    int type;
+    double sample_rate;
+};
+
+static void pack_string(msgpack_packer *ppck, char *str)
+{
+    int len = strlen(str);
```

Comment:
for incoming strings that you already know it lengths, please avoid strlen(). You could add a third parameter to this function to specify the length, if length comes as -1 you can do strlen(), otherwise use the given value

Reply:
> You could add a third parameter to this function to specify the length, if length comes as -1 you can do strlen(), otherwise use the given value

Fixed the interface of `pack_string()` as per your suggestion.